engine/source_sym_engine.ml

```diff
@@ -0,0 +1,175 @@
+open Ast
+
+type constraint_tree =
+  | Empty
+  | Leaf of source_expr
+  | Node of (constructor * constraint_tree) list * constraint_tree
+
+let string_of_source_expr = function
+  | SBlackbox s -> s
+
+let print_result stree =
+  let bprintf = Printf.bprintf
+  in
+  let bprint_source_expr buf = function SBlackbox s -> bprintf buf "%s" s in
+  let rec bprint_list ~sep bprint buf = function
+    | [] -> ()
+    | [x] -> bprint buf x
+    | x :: xs ->
+       bprintf buf "%a%t%a"
+         bprint x
+         sep
+         (bprint_list ~sep bprint) xs in
+  let break ntabs buf =
+    bprintf buf "\n%s" (BatList.init ntabs (fun _ -> "\t") |> String.concat "") in
+  let rec bprint_tree ntabs buf tree =
+    let sep = break (ntabs+1) in
+    let bprint_constructor buf k = match k with
+      | Variant s -> bprintf buf "Variant %s" s
+      | Int i -> bprintf buf "Int %d" i
+      | Bool b -> bprintf buf "Bool %b" b
+      | String s -> bprintf buf "String \"%s\"" s
+      | Tuple narity -> bprintf buf "Tuple[%d]" narity
+      | Nil ->  bprintf buf "Nil"
+      | Cons -> bprintf buf "Cons"
+    in
+    match tree with
+    | Empty -> bprintf buf "Empty"
+    | Leaf expr ->
+      bprintf buf
+        "Leaf='%a'"
+        bprint_source_expr expr
+    | Node (k_cst_list, fallback_cst) ->
+      bprintf buf "Node:{\
+                   %a \
+                   %t} Fallback: %a"
+        (bprint_list ~sep:sep
+           (fun buf (k,cst) -> bprintf buf "%t%a -> %t%a"
+               sep
+               bprint_constructor k
+               (break (ntabs+2))
+               (bprint_tree (ntabs+1)) cst))
+        k_cst_list
+        (break ntabs)
+        (bprint_tree (ntabs+1)) fallback_cst
+  in
+  let buf = Buffer.create 42 in
+  bprint_tree 0 buf stree;
+  BatIO.write_line BatIO.stdout (Buffer.contents buf)
+
+type row = pattern list * source_expr
+
+type hashable_constructor = | HConstructor of constructor | HWildcard
+
+let sym_exec source =
+  let group_constructors rows : (constructor * row list) list * row list =
+    let grouptbl = Hashtbl.create 42
+    in
+    let (ksttbl: (source_blackbox , int) Hashtbl.t) = Hashtbl.create 42
```

Comment:
`source_blackbox` is unfortunate here, you probably mean `string` as the payload of `Variant` constructors, right? (This could be a `type variant_constructor = string` alias that we would use in the AST definition.)